### DISCUSSION TITLE
Errors sending data to Orion Broker with fipy library (AIGreenWaste project)

### DISCUSSION THREAD

**Discussion by cmelchiorre** [c0c0n3/kitt4sme.live#165] [Q&A] [answered]
Hello Kitt4SME tech teams!
 
I am trying to implement an agent to provide biodigester-related entities to the Orion context broker: To do this, I'm replicating the code in the sample Roughnator service and using your fipy library. It seems pretty simple but I'm getting an error when performing the upsert to the context broker and can't understand where the problem is.
Below is all the relevant information and the log of the error from the context broker:
 
This is the entity (DailyProtocolUpdate) definition (there are >30 properties defined, for shortness I include only a few):

	class DailyProtocolUpdate(BaseEntity):
	    
	        type = 'DailyProtocolUpdate'
	    
	        ore_servizio__agitatore_digestore: Optional[FloatAttr]
	        ore_servizio__carico: Optional[FloatAttr]
	        ore_servizio__agitatore_digestore_sovraccarico: Optional[FloatAttr]
	        #
	        .
	        . 
	        .
	    
	        def toJSON(self):
	            return json.dumps(self, default=lambda o: o.__dict__, 
	                sort_keys=True, indent=4)

This is the main.py code:

	@app.post("/update")
	async def post_updates( dailyProtocol: DailyProtocolUpdate,
	             fiware_service: Optional[str] = Header(None),
	             fiware_servicepath: Optional[str] = Header(None),
	             fiware_correlator: Optional[str] = Header(None) ):
	
		dailyProtocol.set_id_with_type_prefix(date.today().strftime("%Y-%m-%d"))
		
		ctx = FiwareContext(
		    service=str(fiware_service), 
		    service_path=str(fiware_servicepath),
		    correlator=str(fiware_correlator)
		)
		
		update_context( ctx, dailyProtocol )
		
		return dailyProtocol
	
	# send daily protocol udate to Orion
	def update_context(ctx: FiwareContext, dailyProtocol: DailyProtocolUpdate ):
		log.processing_updates( ctx, dailyProtocol )
		orion = OrionClient(config.orion_base_url(), ctx)
		orion.upsert_entity(dailyProtocol)
	
	def _format_mgs(lines: List[Any]) -> str:
		ls = [f"{line}\n" for line in lines]
		return ''.join(ls)
	
	def processing_updates(ctx: FiwareContext, dailyProtocol: DailyProtocolUpdate):
		header = f"going to process updates for {ctx}:"
		msg = _format_mgs([header] + [ dailyProtocol.toJSON() ])
		print(msg)

This is the curl command with which I invoke the REST service update endpoint:

    curl -v -s -d '{
        "ore_servizio__carico": { "type": "Number", "value": 1.0 },
        "ore_servizio__agitatore_digestore": { "type": "Number", "value": 1.0 },
        "ore_servizio__agitatore_digestore_sovraccarico": { "type": "Number", "value": 1.0 },
        .
        .
        .        
        "type": "DailyProtocolUpdate",
      }' \
    -H 'Content-Type: application/json; charset=utf-8"' -X POST http://localhost:8000/update | jq

And finally, the following is the output I get on the service's console:

    going to process updates for service='None' service_path='None' correlator='None':
    {
        "energia__consumo_energia": {
            "type": "Number",
            "value": 1.0
        },
        "energia__energia_cogenerazione": {
            "type": "Number",
            "value": 1.0
        },
        "energia__energia_termica_digestore": {
            "type": "Number",
            "value": 1.0
        },
        .
        .
        .
        "id": "urn:ngsi-ld:DailyProtocolUpdate:2022-08-09",
        "type": "DailyProtocolUpdate"
    }
    
    INFO:     172.18.0.1:34838 - "POST /update2 HTTP/1.1" 500 Internal Server Error
    ERROR:    Exception in ASGI application
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi
        result = await app(self.scope, self.receive, self.send)
      File "/usr/local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
        return await self.app(scope, receive, send)
      File "/usr/local/lib/python3.8/site-packages/fastapi/applications.py", line 212, in __call__
        await super().__call__(scope, receive, send)
      File "/usr/local/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
        await self.middleware_stack(scope, receive, send)
      File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
        raise exc
      File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
        await self.app(scope, receive, _send)
      File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
        raise exc
      File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
        await self.app(scope, receive, sender)
      File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 656, in __call__
        await route.handle(scope, receive, send)
      File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 259, in handle
        await self.app(scope, receive, send)
      File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 61, in app
        response = await func(request)
      File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 226, in app
        raw_response = await run_endpoint_function(
      File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 159, in run_endpoint_function
        return await dependant.call(**values)
      File "/src/./aigreen_ctxprovider/main.py", line 67, in post_updates_2
        update_context( ctx, dailyProtocol )
      File "/src/./aigreen_ctxprovider/main.py", line 76, in update_context
        orion.upsert_entity(dailyProtocol)
      File "/usr/local/lib/python3.8/site-packages/fipy/ngsi/orion.py", line 46, in upsert_entity
        self._http.post(url=url, json_payload=data.dict(),
      File "/usr/local/lib/python3.8/site-packages/fipy/http/jclient.py", line 72, in post
        return self._handle_response(response)
      File "/usr/local/lib/python3.8/site-packages/fipy/http/jclient.py", line 26, in _handle_response
        r.raise_for_status()
      File "/usr/local/lib/python3.8/site-packages/requests/models.py", line 960, in raise_for_status
        raise HTTPError(http_error_msg, response=self)
    requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http://orion:1026/v2/entities?options=upsert

It seems that the entity is correctly parsed (from the log output), but when it is passed to orion.upsert_entity() it produces an error.

**Question 1**: What am I doing wrong?

**Question 2**: I would like to be able to pass to the POST a simpler body, like the following one (ie. a simple name/value JSON format), and then have the post_updates function perform internally the conversion to an NGSI-LD entity. There's a way I can do this with the fipy library? 

    curl -v -s -d '{
             "ore_servizio__carico": 1.0,
             "ore_servizio__agitatore_digestore": 1.0,
             "ore_servizio__agitatore_digestore_sovraccarico": 1.0,
             .
             .
             .
             "type": "DailyProtocolUpdate",
      }' \
    -H 'Content-Type: application/json; charset=utf-8"' -X POST http://localhost:8000/testupdate | jq

I guess that one option would be to define a second class DailyProtocolUpdateRaw(BaseModel) with fields as `: float = 0.0` then making a manual assignment field by field, but due to the large number of properties I wonder if there is a more "automatic" way to achieve this...

**Comment by c0c0n3** [comment]
hi @cmelchiorre :-)

### Question 1
It looks like Orion isn't happy with the upsert call. It could be there's a translation glitch in the service where the original JSON gets converted to something Orion can't digest. Can you try sending the original JSON you input to the service directly to Orion? It seems you already have it in NGSI v2 format, so it should work---the only thing that's missing is the ID prop it seems, but you could add that manually.

```console
curl -v -s -d '{
    "ore_servizio__carico": { "type": "Number", "value": 1.0 },
    "ore_servizio__agitatore_digestore": { "type": "Number", "value": 1.0 },
    "ore_servizio__agitatore_digestore_sovraccarico": { "type": "Number", "value": 1.0 },
    .
    .
    .        
    "type": "DailyProtocolUpdate",
  }' \
-H 'Content-Type: application/json; charset=utf-8"' -X POST http://orion:1026/v2/entities?options=upsert
```

Can you copy-paste here the exact curl request you make as well as Orion's response?


### Question 2
FiPy doesn't support any automatic conversion at the moment. The reason is that there are cases where there's many way to map a JSON doc to an NGSI entity---e.g. should a JSON field of "x: 1" become a Number, Integer or even Bool prop? On the other hand, there are many situations where what you want is just a straightforward conversion from JSON, so that a JSON number gets mapped to an NGSI Number prop, a string to a Text prop, an object to a StructuredValue, etc. This can actually be done automatically, if other open call devs think this could be useful to them too, please open an issue at the FiPy repo, we'll implement it :-)

**Reply by cmelchiorre** [reply]
Hi, thanks for your reply!

### Question 1

If I execute the curl command as below,

	curl -v -s -d '{
	    "energia__consumo_energia": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__energia_cogenerazione": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__energia_termica_digestore": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__media_acido_solfidrico_h25": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__media_biossido_carbonio_co2": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__media_metano_ch4": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__media_ossigeno_o2": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__potere_calorifico_inferiore": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "energia__produzione_biogas": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__acqua_fresca": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__acqua_umidificazione": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__corse_pistone_inoculazione": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__corse_pistone_scarico": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__materiale_fresco": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__materiale_strutturale": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__quantita_carico_verde": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__quantita_materiale_mescolatore": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__quantita_materiale_tiger": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "input_output__quantita_scarico_mescolatore": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__agitatore_digestore": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__agitatore_digestore_sovraccarico": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__carico": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__cogenerazione": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__mescolatore": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__tiger": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__torcia_biogas_livello_1": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "ore_servizio__torcia_biogas_livello_2": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "temperatura_carico_livelli__livello_digestore": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "temperatura_carico_livelli__max_quantita_agitatore_giornaliero": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "temperatura_carico_livelli__temperatura_digestore_carico": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "temperatura_carico_livelli__temperatura_digestore_centro": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "temperatura_carico_livelli__temperatura_digestore_scarico": {
	        "type": "Number",
	        "value": 1.0
	    },
	    "id": "urn:ngsi-ld:DailyProtocolUpdate:2022-09-19",
	    "type": "DailyProtocolUpdate"
	  }' \
	-H 'Content-Type: application/json; charset=utf-8"' -X POST http://192.168.33.100:1026/v2/entities?options=upsert
	
This is the response I get:

	*   Trying 192.168.33.100:1026...
	* TCP_NODELAY set
	* Connected to 192.168.33.100 (192.168.33.100) port 1026 (#0)
	> POST /v2/entities?options=upsert HTTP/1.1
	> Host: 192.168.33.100:1026
	> User-Agent: curl/7.68.0
	> Accept: */*
	> Content-Type: application/json; charset=utf-8"
	> Content-Length: 3294
	> Expect: 100-continue
	>
	* Mark bundle as not supporting multiuse
	< HTTP/1.1 100 Continue
	* We are completely uploaded and fine
	* Mark bundle as not supporting multiuse
	< HTTP/1.1 204 No Content
	< Connection: Keep-Alive
	< Location: /v2/entities/urn:ngsi-ld:DailyProtocolUpdate:2022-09-05?type=DailyProtocolUpdate
	< Fiware-Correlator: 06eb81a0-3815-11ed-b59f-0242ac120006
	< Date: Mon, 19 Sep 2022 12:17:27 GMT
	<
	* Connection #0 to host 192.168.33.100 left intact
	
It looks like it worked... I also get the Json back if I then call

	curl -v -s -X  GET http://192.168.33.100:1026/v2/entities/urn:ngsi-ld:DailyProtocolUpdate:2022-09-05?type=DailyProtocolUpdate

On the other hand, on my python application's log (the one using the FiPy calls), I get the following error trace

	INFO:     172.18.0.8:42610 - "POST /update HTTP/1.1" 500 Internal Server Error
	ERROR:    Exception in ASGI application
	Traceback (most recent call last):
	  File "/usr/local/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi
	    result = await app(self.scope, self.receive, self.send)
	  File "/usr/local/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
	    return await self.app(scope, receive, send)
	  File "/usr/local/lib/python3.8/site-packages/fastapi/applications.py", line 212, in __call__
	    await super().__call__(scope, receive, send)
	  File "/usr/local/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
	    await self.middleware_stack(scope, receive, send)
	  File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
	    raise exc
	  File "/usr/local/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
	    await self.app(scope, receive, _send)
	  File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
	    raise exc
	  File "/usr/local/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
	    await self.app(scope, receive, sender)
	  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 656, in __call__
	    await route.handle(scope, receive, send)
	  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 259, in handle
	    await self.app(scope, receive, send)
	  File "/usr/local/lib/python3.8/site-packages/starlette/routing.py", line 61, in app
	    response = await func(request)
	  File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 226, in app
	    raw_response = await run_endpoint_function(
	  File "/usr/local/lib/python3.8/site-packages/fastapi/routing.py", line 159, in run_endpoint_function
	    return await dependant.call(**values)
	  File "/src/./aigreen_ctxprovider/main.py", line 48, in post_updates
	    update_context( ctx, dailyProtocol )
	  File "/src/./aigreen_ctxprovider/main.py", line 76, in update_context
	    orion.upsert_entity(dailyProtocol)
	  File "/usr/local/lib/python3.8/site-packages/fipy/ngsi/orion.py", line 46, in upsert_entity
	    self._http.post(url=url, json_payload=data.dict(),
	  File "/usr/local/lib/python3.8/site-packages/fipy/http/jclient.py", line 72, in post
	    return self._handle_response(response)
	  File "/usr/local/lib/python3.8/site-packages/fipy/http/jclient.py", line 26, in _handle_response
	    r.raise_for_status()
	  File "/usr/local/lib/python3.8/site-packages/requests/models.py", line 960, in raise_for_status
	    raise HTTPError(http_error_msg, response=self)
	requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http://orion:1026/v2/entities?options=upsert
	
Note that the json provided as body to the first curl command is directly taken from the log's output produced by printing dailyProtocol.toJSON() in the into() methond by the following code, which should correspond to what is passed to the OrionClient via `orion.upsert_entity`

	# main.py
	
	def update_context(ctx: FiwareContext, dailyProtocol: DailyProtocolUpdate ):
	    log.processing_updates( ctx, dailyProtocol )
	    orion = OrionClient(config.orion_base_url(), ctx)
	    orion.upsert_entity(dailyProtocol)

	# log.py
	
	def processing_updates(ctx: FiwareContext, dailyProtocol: DailyProtocolUpdate):
	    header = f"going to process updates for {ctx}:"
	    msg = _format_mgs([header] + [ dailyProtocol.toJSON() ])
	    info(msg)

	def info(msg: str):
	    print(msg)
	    
	
	# ngsy.py	
	
	class DailyProtocolUpdate(BaseEntity):

		type = 'DailyProtocolUpdate'
		
		ore_servizio__agitatore_digestore: Optional[FloatAttr]
		ore_servizio__carico: Optional[FloatAttr]
		ore_servizio__agitatore_digestore_sovraccarico: Optional[FloatAttr]
   		.
   		. definition of the other entity's fields
   		.
	
		def toJSON(self):
			return json.dumps(self, default=lambda o: o.__dict__, 
			    sort_keys=True, indent=4)


### Question 2

Thanks for the suggestion :) since I consider this a very useful function and would be happy to reuse your FiPy library also in future projects (and, why not, the whole Kitt4SME platform) I will surely open an issue on the FiPy repo, hoping others will find it usefull too!  Thanks.

**Reply by c0c0n3** [reply]
Hi @cmelchiorre :-)

Thanks for the very detailed debug info, it really helps!

### Question 1
Is it possible that the URL used in your Python code

- http://orion:1026/v2/entities?options=upsert

is not the same as the one you use in the `curl` command

- http://192.168.33.100:1026/v2/entities?options=upsert

i.e. what IP address does `orion` resolve to? Keep in mind if you have a Docker compose env w/ an `orion` service and you're running your script outside of the Docker compose env, then `http://orion:1026` won't work...

On a side note, Pydantic comes with a `json` method you can use to convert a model to JSON and we also have a `to_json` method in `BaseEntity` that won't output empty optional NGSI attributes

- https://github.com/c0c0n3/kitt4sme.fipy/blob/main/fipy/ngsi/entity.py#L123

There's some examples at the top of the file.

### Question 2
I agree w/ you, automatic JSON to NGSI translation could come in handy for others too, so we should add it to FiPy. If you [open an issue for a feature request](https://github.com/c0c0n3/kitt4sme.fipy/issues), we'll add it to our project plan and then assign it to someone to work on.

**Comment by cmelchiorre** [comment]
Hi @c0c0n3

The python script with the fipy calls runs in its own container by the same docker-compose file, where the orion name is resolved correctly. 
I can verify that the message sent by the fipy code reaches the orion broker (which just then returns an error) by the following message from the orion container's log:

    time=Wednesday 21 Sep 06:51:45 2022.538Z | lvl=INFO | corr=None | trans=1663742053-924-00000000016 | from=172.18.0.8 | 
    srv=pending | subsrv=pending | comp=Orion | op=rest.cpp[1094]:servicePathSplit | msg=Service Path 0: 'None'

I also made the following test: I `docker exec`'d inside the python container and from there executed a curl with the URL http://orion:1026/... and again the curl worked.

Finally, I replaced my custom toJSON() method with the call to the to_json() base class method, but repeating all the tests nothing changes (the generated json also looks the same).

Could the issue be somehow related to the `msg=Service Path 0: 'None'` message in the orion log?

**Comment by cmelchiorre** [comment, marked as answer]
I fixed it by setting default values for `FIWARE-Service` and `FIWARE-ServicePath` headers in the python code:

    @app.post("/update")
    async def post_updates( dailyProtocol: DailyProtocolUpdate,
                     fiware_service: Optional[str] = Header("default"),
                     fiware_servicepath: Optional[str] = Header("/",
                     fiware_correlator: Optional[str] = Header(None) ):
    .
    .
    .
